generate.go

```diff
@@ -0,0 +1,179 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package terramate
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/hashicorp/hcl/v2/hclsyntax"
+	"github.com/hashicorp/hcl/v2/hclwrite"
+	"github.com/madlambda/spells/errutil"
+	"github.com/mineiros-io/terramate/hcl"
+)
+
+const (
+	// GeneratedTfFilename is the name of the terramate generated tf file.
+	GeneratedTfFilename = "_gen_terramate.tm.tf"
+
+	// GeneratedCodeHeader is the header added on all generated files.
+	GeneratedCodeHeader = "// GENERATED BY TERRAMATE: DO NOT EDIT\n\n"
+)
+
+// Generate will walk all the directories starting from basedir generating
+// code for any stack it finds as it goes along
+//
+// It will return an error if it finds any invalid terramate configuration files
+// of if it can't generate the files properly for some reason.
+//
+// The provided basedir must be an absolute path to a directory.
+func Generate(basedir string) error {
+	if !filepath.IsAbs(basedir) {
+		return fmt.Errorf("Generate(%q): basedir must be an absolute path", basedir)
+	}
+
+	info, err := os.Lstat(basedir)
+	if err != nil {
+		return fmt.Errorf("Generate(%q): checking basedir: %v", basedir, err)
+	}
+
+	if !info.IsDir() {
+		return fmt.Errorf("Generate(%q): basedir is not a directory", basedir)
+	}
+
+	stacks, err := ListStacks(basedir)
+	if err != nil {
+		return fmt.Errorf("Generate(%q): listing stack: %v", basedir, err)
+	}
+
+	var errs []error
+
+	for _, stack := range stacks {
+		// At the time the most intuitive way was to start from the stack
+		// and go up until reaching the basedir, looking for a config.
+		// Basically navigating from the order of precedence, since
+		// more specific configuration overrides base configuration.
+		// Not the most optimized way (re-parsing), we can improve later
+		tfcode, err := generateStackConfig(basedir, stack.Dir)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		if tfcode == nil {
+			continue
+		}
+
+		genfile := filepath.Join(stack.Dir, GeneratedTfFilename)
+		errs = append(errs, os.WriteFile(genfile, tfcode, 0666))
+	}
+
+	if err := errutil.Chain(errs...); err != nil {
+		return fmt.Errorf("Generate(%q): %w", basedir, err)
+	}
+
+	return nil
+}
+
+func generateStackConfig(basedir string, configdir string) ([]byte, error) {
+	if !strings.HasPrefix(configdir, basedir) {
+		// check if we are outside of basedir, time to stop
+		return nil, nil
+	}
+
+	configfile := filepath.Join(configdir, ConfigFilename)
+
+	if _, err := os.Stat(configfile); err != nil {
```

Comment:
maybe check for `os.IsNotExists(err)` and return the error if the file exists but stat failed whatever the other reason.